sematic/log_reader.py

```diff
@@ -0,0 +1,429 @@
+        return Cursor(
+            source_log_key=None,
+            source_file_line_index=-1,
+            filter_strings=filter_strings,
+            run_id=run_id,
+        )
+
+
+@dataclass
+class LogLine:
+    source_file: str
+    source_file_index: int
+    line: str
+
+
+def load_log_lines(
+    run_id: str,
+    continuation_cursor: Optional[str],
+    max_lines: int,
+    filter_strings: Optional[List[str]] = None,
+) -> LogLineResult:
+    """Load a portion of the logs for a particular run
+
+    Parameters
+    ----------
+    run_id:
+        The id of the run to get logs for
+    continuation_cursor:
+        A cursor indicating where to continue reading logs from. Should be
+        None if the logs are being read from the beginning.
+    max_lines:
+        The highest number of log lines that should be returned at once
+    filter_strings:
+        Only log lines that contain ALL of the strings in this list will
+        be included in the result
+
+    Returns
+    -------
+    A subset of the logs for the given run
+    """
+    run = get_run(run_id)
+    run_state = FutureState[run.future_state]  # type: ignore
+    still_running = run_state.is_terminal() or run_state == FutureState.RAN
+    resolution = get_resolution(run.root_id)
+    filter_strings = filter_strings if filter_strings is not None else []
+    cursor = (
+        Cursor.from_token(continuation_cursor)
+        if continuation_cursor is not None
+        else Cursor.nothing_found(filter_strings, run_id)
+    )
+    if cursor.run_id != run_id:
+        raise ValueError(
+            f"Tried to continue a log search of {run_id} using a "
+            f"continuation cursor from {cursor.run_id}"
+        )
+    if set(cursor.filter_strings) != set(filter_strings):
+        raise ValueError(
+            f"Tried to continue a log search of {run_id} using a "
+            f"different set of filters than were used in the cursor."
+        )
+
+    if ResolutionStatus[resolution.status] in (  # type: ignore
+        ResolutionStatus.CREATED,
+        ResolutionStatus.SCHEDULED,
+    ):
+        return LogLineResult(
+            more_before=False,
+            more_after=True,
+            lines=[],
+            continuation_cursor=cursor.to_token(),
+            log_unavaiable_reason="Resolution has not started yet.",
+        )
+    filter_strings = filter_strings if filter_strings is not None else []
+    if FutureState[run.future_state] == FutureState.CREATED:  # type: ignore
+        return LogLineResult(
+            more_before=False,
+            more_after=True,
+            lines=[],
+            continuation_cursor=cursor.to_token(),
+            log_unavaiable_reason="The run has not yet started executing.",
+        )
+    # looking for external jobs to determine inline is only valid
+    # since we know the run has at least reached SCHEDULED due to it
+    # not being CREATED.
+    is_inline = len(run.external_jobs) == 0
+    if is_inline:
```

Comment:
We may want to persist this. It would even be nice to show it in the UI. Separate PR obviously.